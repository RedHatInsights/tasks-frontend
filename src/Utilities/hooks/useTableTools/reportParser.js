// extended utilities parse jobs with report_json available

export const encodeCsvCell = (
  cell // https://www.ietf.org/rfc/rfc4180.txt
) =>
  typeof cell === 'string'
    ? `"${cell.replaceAll('"', '""')}"`
    : Array.isArray(cell)
    ? `"${cell.map(encodeCsvCell)}"`
    : cell;

const parseReportDiagnosis = (detail) => {
  const { diagnosis } = detail || {};

  return diagnosis?.context ?? diagnosis?.[0]?.context ?? '';
};

const parseReportRemediations = (detail) => {
  const { remediations } = detail || {};

  if (Array.isArray(remediations)) {
    return remediations.map(({ type, context }) => ({ type, context }));
  }

  if (typeof remediations === 'object') {
    return [
      { type: remediations?.type ?? '', context: remediations?.context ?? '' },
    ];
  }

  return [{ type: '', context: '' }];
};

const parseReportJson = (report) => {
  const { entries } = report;

  if (entries === undefined || entries === null) {
    return [];
  }

  return entries
    .map(({ title, severity, key, summary, detail }) => {
      const diagnosis = parseReportDiagnosis(detail);
      const remediations = parseReportRemediations(detail);

      if (remediations.length > 0) {
        return remediations.map((remediation) => ({
          title,
          severity,
          key,
          summary,
          diagnosis,
          remediationType: remediation.type,
          remediationContext: remediation.context,
        }));
      } else {
        return [
          {
            title,
            severity,
            key,
            summary,
            diagnosis,
            remediationType: '',
            remediationContext: '',
          },
        ];
      }
    })
    .flat();
};

export const parseJobReport = (job) => {
  const { report_json } = job?.results || {};

  if (report_json !== undefined && report_json !== null) {
    const issues = parseReportJson(report_json);

    if (issues.length === 0) {
      return [job];
    } else {
      return issues.map((issue) => ({ ...job, issue_parsed: issue }));
    }
  }

  return [job];
};

export const prepareItems = (items) =>
  items.map((job) => parseJobReport(job)).flat();
