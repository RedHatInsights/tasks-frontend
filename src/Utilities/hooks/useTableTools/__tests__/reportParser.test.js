import cloneDeep from 'lodash/cloneDeep';
import { parseJobReport, prepareItems } from '../reportParser';
import {
  fixturesExtendedReport,
  fixturesPlainReport,
} from './__fixtures__/jobResultsItems.fixtures';

describe('parseJobReport', () => {
  it('should parse a job without report', () => {
    const parsed = parseJobReport(fixturesPlainReport);

    expect(parsed).toHaveLength(1);
    expect(parsed).toMatchSnapshot();
  });

  it('should have correct length of parsed report_json', () => {
    const parsed = parseJobReport(fixturesExtendedReport);

    expect(parsed).toHaveLength(8);
  });

  it('should add issue_parsed', () => {
    const parsed = parseJobReport(fixturesExtendedReport);

    expect(parsed.map(({ issue_parsed }) => issue_parsed)).toMatchSnapshot();
  });

  it('parses multiple remediatios for the extended report', () => {
    const fixtures = cloneDeep(fixturesExtendedReport);
    fixtures.results.report_json.entries[0].detail?.remediations?.push({
      type: 'info',
      context: 'Some info remediation.',
    });
    const result = parseJobReport(fixtures);

    expect(result).toHaveLength(9);
    expect(
      result.some(
        ({ issue_parsed }) =>
          issue_parsed.remediationContext === 'Some info remediation.'
      )
    ).toBeTruthy();
  });

  it('parses issue diagnosis for the extended report', () => {
    const fixtures = cloneDeep(fixturesExtendedReport);
    fixtures.results.report_json.entries[0].detail?.remediations?.push({
      type: 'info',
      context: 'Some info remediation.',
    });
    fixtures.results.report_json.entries[0].detail = {
      ...fixtures.results.report_json.entries[0].detail,
      diagnosis: { context: 'Test diagnosis text.' },
    };
    const result = parseJobReport(fixtures);

    expect(result).toHaveLength(9);
    expect(
      result.filter(
        ({ issue_parsed }) => issue_parsed.diagnosis === 'Test diagnosis text.'
      )
    ).toHaveLength(2);
  });
});

describe('prepareItems', () => {
  it('should prepare items with report_json', () => {
    const parsed = prepareItems([fixturesPlainReport, fixturesExtendedReport]);

    expect(parsed).toHaveLength(9);
  });
});
