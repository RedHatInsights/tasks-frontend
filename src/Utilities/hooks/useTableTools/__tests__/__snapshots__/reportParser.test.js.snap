// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseJobReport should add issue_parsed 1`] = `
[
  {
    "diagnosis": "",
    "key": "1b91322362ae7830e48eee7811be9527747de8",
    "remediationContext": "Lorem ipsum dolor sit amet, "consectetuer" adipiscing elit.",
    "remediationType": "hint",
    "severity": "info",
    "summary": "The following repositories are not supported by Red Hat and are excluded from the list of repositories used during the upgrade.
- codeready-builder-for-rhel-9-x86_64-eus-rpms
- codeready-builder-for-rhel-9-aarch64-rpms
- rhui-codeready-builder-for-rhel-9-aarch64-rhui-rpms
- codeready-builder-for-rhel-9-s390x-eus-rpms
- rhui-codeready-builder-for-rhel-9-x86_64-rhui-rpms
- codeready-builder-for-rhel-9-aarch64-eus-rpms
- codeready-builder-for-rhel-9-s390x-rpms
- codeready-builder-beta-for-rhel-9-aarch64-rpms
- codeready-builder-for-rhel-9-x86_64-rhui-rpms
- codeready-builder-beta-for-rhel-9-s390x-rpms
- codeready-builder-beta-for-rhel-9-x86_64-rpms
- codeready-builder-for-rhel-9-ppc64le-eus-rpms
- codeready-builder-for-rhel-9-ppc64le-rpms
- codeready-builder-beta-for-rhel-9-ppc64le-rpms
- codeready-builder-for-rhel-9-x86_64-rpms
- codeready-builder-for-rhel-9-rhui-rpms",
    "title": "Excluded target system repositories",
  },
  {
    "diagnosis": "",
    "key": "1c7a949a747ec9890f04bf4321de7280970715",
    "remediationContext": "",
    "remediationType": "",
    "severity": "inhibitor",
    "summary": "The supported OS releases for the upgrade process:
 RHEL 8.8
RHEL 8.10
RHEL-SAPHANA 8.8
RHEL-SAPHANA 8.10",
    "title": "The installed OS version is not supported for the in-place upgrade to the target RHEL version",
  },
  {
    "diagnosis": "",
    "key": "ac7030e05de248d34f08a9fa040b352bc410a3",
    "remediationContext": "",
    "remediationType": "",
    "severity": "high",
    "summary": "On legacy (BIOS) systems, GRUB2 core (located in the gap between the MBR and the first partition) cannot be updated during the rpm transaction and Leapp has to initiate the update running "grub2-install" after the transaction. No action is needed before the upgrade. After the upgrade, it is recommended to check the GRUB configuration.",
    "title": "GRUB2 core will be automatically updated during the upgrade",
  },
  {
    "diagnosis": "",
    "key": "8fb818f8413bd617c2a55b69b8e10ff03d7c72",
    "remediationContext": "",
    "remediationType": "",
    "severity": "info",
    "summary": "SElinux relabeling will be scheduled as the status is permissive/enforcing.",
    "title": "SElinux relabeling will be scheduled",
  },
  {
    "diagnosis": "",
    "key": "39d7183dafba79a4bbb1e70b0ef2bbe5b1772f",
    "remediationContext": "Make sure there are no SElinux related warnings after the upgrade and enable SElinux manually afterwards. Notice: You can ignore the "/root/tmp_leapp_py3" SElinux warnings.",
    "remediationType": "hint",
    "severity": "low",
    "summary": "SElinux will be set to permissive mode. Current mode: enforcing. This action is required by the upgrade process to make sure the upgraded system can boot without beinig blocked by SElinux rules.",
    "title": "SElinux will be set to permissive mode",
  },
  {
    "diagnosis": "",
    "key": "e738f78bc8f3a84414210e3b609057139d1855",
    "remediationContext": "If you depend on remote root logins using passwords, consider setting up a different user for remote administration. Otherwise you can ignore this message.",
    "remediationType": "hint",
    "severity": "high",
    "summary": "RHEL9 no longer allows remote root logins, but the server configuration explicitly overrides this default. The configuration file will not be updated and root is still going to be allowed to login with password. This is not recommended and considered as a security risk.",
    "title": "Remote root logins globally allowed using password",
  },
  {
    "diagnosis": "",
    "key": "96da6937c25c6492e4f1228e46795989fd3718",
    "remediationContext": "",
    "remediationType": "",
    "severity": "info",
    "summary": "OpenSSH server configuration needs to be modified to contain Include directive for the RHEL9 to work properly and integrate with the other parts of the OS. The following snippet will be added to the /etc/ssh/sshd_config during the ApplicationsPhase: \`Include /etc/ssh/sshd_config.d/*.conf\`",
    "title": "The upgrade will prepend the Include directive to OpenSSH sshd_config",
  },
  {
    "diagnosis": "",
    "key": "693963253195f41852045b6d630a1f4c7a193d",
    "remediationContext": "",
    "remediationType": "",
    "severity": "info",
    "summary": "After the upgrade, this system will be automatically registered into Red Hat Insights. To skip the automatic registration, use the '--no-insights-register' command line option or set the LEAPP_NO_INSIGHTS_REGISTER environment variable.",
    "title": "Automatic registration into Red Hat Insights",
  },
]
`;

exports[`parseJobReport should parse a job without report 1`] = `
[
  {
    "display_name": "host_abc",
    "results": {
      "message": "Lorem ipsum dolor sit amet",
      "report": "Lorem ipsum dolor sit amet, "consectetuer" adipiscing elit. Maecenas fermentum, sem in pharetra pellentesque, velit turpis > !volutpat ante, in pharetra metus odio a lectus. ",
    },
    "status": "Success",
  },
]
`;
